hack/k-license/k-license.go

```diff
@@ -0,0 +1,151 @@
+	helpFlag     = flag.Bool("help", false, "Display this help message and exit.")
+)
+
+func main() {
+	flag.Parse()
+	if *helpFlag {
+		printHelp(os.Args[0])
+		return
+	}
+	excludeList := strings.Split(*excludeDirs, ",")
+	err := filepath.Walk(*path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() && contains(excludeList, info.Name()) {
+			return filepath.SkipDir
+		}
+		if !info.IsDir() && (isCodeFile(path) || isBuildFile(path)) && !hasLicense(path) {
+			currentYear := strconv.Itoa(time.Now().Year())
+			err := addLicense(path, *templatesDir, currentYear)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("Modified %s file\n", path)
+		}
+
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func contains(list []string, str string) bool {
+	for _, item := range list {
+		if item == str {
+			return true
+		}
+	}
+	return false
+}
+
+func printHelp(progName string) {
+	fmt.Fprintf(os.Stderr, "Usage: %s [--help] [-d dir] [--exclude ] [--templates]\n", progName)
+	fmt.Fprintln(os.Stderr, "Add a copyright/license header to all source files in a Git repository.\n-")
+	fmt.Fprintln(os.Stderr, "Options:")
+	flag.PrintDefaults()
+	fmt.Fprintln(os.Stderr, "\nExamples:")
+	fmt.Fprintf(os.Stderr, "  %s               # Add license headers to all source files in the current directory and its subdirectories, using the default file extensions and templates directory.\n", progName)
+	fmt.Fprintf(os.Stderr, "  %s --dir /path/to/repo --exclude vendor,node_modules --templates /path/to/templates   # Add license headers to all .go, .sh, .py, makefile, docker files in /path/to/repo and its subdirectories, using the license header templates in /path/to/templates.\n", progName)
+}
+
+func isCodeFile(path string) bool {
+	ext := strings.ToLower(filepath.Ext(path))
+	switch ext {
+	case ".go", ".c", ".h", ".cpp", ".py", ".ipynb", ".java":
+		return true
+	}
+	return false
+}
+func isBuildFile(path string) bool {
+	switch {
+	case strings.HasSuffix(path, ".sh"):
+		return true
+	case strings.HasSuffix(path, "Makefile"):
+		return true
+	case strings.HasSuffix(path, "Dockerfile"):
+		return true
+	}
+	return false
+}
+
+func hasLicense(path string) bool {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return false
+	}
+	return strings.Contains(string(data), "Copyright")
+}
+func getFileType(path string) string {
+	var tmplFilename string
+	switch {
+	case strings.HasSuffix(path, ".sh"):
+		tmplFilename = "boilerplate.sh.txt"
+	case strings.HasSuffix(path, "Makefile"):
+		tmplFilename = "boilerplate.Makefile.txt"
+	case strings.HasSuffix(path, "Dockerfile"):
+		tmplFilename = "boilerplate.Dockerfile.txt"
+	case filepath.Ext(path) == ".py":
+		tmplFilename = "boilerplate.py.txt"
+	case filepath.Ext(path) == ".go":
+		tmplFilename = "boilerplate.go.txt"
+	default:
+		tmplFilename = "boilerplate.tf.txt"
+	}
+	return tmplFilename
+}
```

Comment:
**Suggestion:** how about storing the different template files in a struct to do the check. It will help resue/reference the template files in other parts of the program (if required later)?

```
type fileTypeTemplate struct {
    extension string
    template  string
}

var fileTypeTemplates = []fileTypeTemplate{
    {".sh", "boilerplate.sh.txt"},
    {"Makefile", "boilerplate.Makefile.txt"},
     ...
}

func getFileType(path string) string {
    for _, ft := range fileTypeTemplates {
        if strings.HasSuffix(path, ft.extension) || filepath.Base(path) == ft.extension {
            return ft.template
        }
    }
    return "boilerplate.tf.txt"
}
```

Reply:
Will it make sense to use everything from struct instead of static templates files. 
since this will be used for generating boilerplates for k,k-sig repos which follows pretty much same templates. 
with this we can avoid --template flag and also creation/copying of templates files.

Reply:
Sorry for being not clear above.

> how about storing the different template files in a struct to do the check. 

In the suggestion above, I meant avoiding the use of hardcoded strings for template file names (such as "boilerplate.sh.txt") and extensions (such as ".sh") within the function. Instead, to define these values in a struct format at the top level to make it easier to reference them throughout the code in the future.

To be clear, we will still be using the existing static template files present here -> https://github.com/kubernetes/org/tree/main/hack/boilerplate.

 I have also included some comments in the code suggestion above to provide further clarity.